### PR TITLE
Mixtapes

### PR DESCRIPTION
**Pull request by dgrosenblatt**
This adds a `GET /mixtape`  endpoint that returns a list of songs based on url parameters. A genre must be specified. A length of time in seconds can optionally be specified as well. A 60-minute mixtape will be returned if no length of time is specified. Mixtapes are not persistent.

Sample request:

```
GET /mixtape?genre=blues&time=2000 HTTP/1.1
Accept: application/vnd.spotify-apprentice-app.com; version=1
```

Sample response:

```
HTTP/1.1 200 OK
{
  "mixtape": {
    "songs": [
      {
        "id": "0c1b633d-0cfa-42b9-9846-1ba3bc144b67",
        "title": "Blues song 2",
        "album": "The blue album",
        "duration": 252,
        "formatted_duration": "4:12",
        "full_title": "The Beatles - Blues song 2 - The blue album",
        "artist_id": "a6f77950-abc2-4eb2-9333-b3924feff571",
        "genre_id": "64ca9927-47e3-4f5e-a93e-49e66965ffff"
      },
      {
        "id": "10804177-cf4b-4131-9e07-f51ba5a10034",
        "title": "Blues song 4",
        "album": "The blue album",
        "duration": 47,
        "formatted_duration": "0:47",
        "full_title": "The Beatles - Blues song 4 - The blue album",
        "artist_id": "a6f77950-abc2-4eb2-9333-b3924feff571",
        "genre_id": "64ca9927-47e3-4f5e-a93e-49e66965ffff"
      },
      {
        "id": "26242bf3-a371-48ef-ae8f-01807820b59d",
        "title": "Blues song 13",
        "album": "The blue album",
        "duration": 306,
        "formatted_duration": "5:6",
        "full_title": "The Beatles - Blues song 13 - The blue album",
        "artist_id": "a6f77950-abc2-4eb2-9333-b3924feff571",
        "genre_id": "64ca9927-47e3-4f5e-a93e-49e66965ffff"
      },
      {
        "id": "17278430-15fe-4220-8da8-76438058836a",
        "title": "Blues song 17",
        "album": "The blue album",
        "duration": 397,
        "formatted_duration": "6:37",
        "full_title": "The Beatles - Blues song 17 - The blue album",
        "artist_id": "a6f77950-abc2-4eb2-9333-b3924feff571",
        "genre_id": "64ca9927-47e3-4f5e-a93e-49e66965ffff"
      },
      {
        "id": "070a340b-bc78-470e-962e-7d31862cb445",
        "title": "Blues song 34",
        "album": "The blue album",
        "duration": 361,
        "formatted_duration": "6:1",
        "full_title": "The Beatles - Blues song 34 - The blue album",
        "artist_id": "a6f77950-abc2-4eb2-9333-b3924feff571",
        "genre_id": "64ca9927-47e3-4f5e-a93e-49e66965ffff"
      },
      {
        "id": "15b43420-dea9-44dd-a0a9-ca13a97f2ac3",
        "title": "Blues song 35",
        "album": "The blue album",
        "duration": 184,
        "formatted_duration": "3:4",
        "full_title": "The Beatles - Blues song 35 - The blue album",
        "artist_id": "a6f77950-abc2-4eb2-9333-b3924feff571",
        "genre_id": "64ca9927-47e3-4f5e-a93e-49e66965ffff"
      },
      {
        "id": "035c8293-7cee-4682-92a2-c3b7654de13c",
        "title": "Blues song 39",
        "album": "The blue album",
        "duration": 263,
        "formatted_duration": "4:23",
        "full_title": "The Beatles - Blues song 39 - The blue album",
        "artist_id": "a6f77950-abc2-4eb2-9333-b3924feff571",
        "genre_id": "64ca9927-47e3-4f5e-a93e-49e66965ffff"
      },
      {
        "id": "16bcd241-fe0e-48db-8539-b6d580bd9f8c",
        "title": "Blues song 43",
        "album": "The blue album",
        "duration": 390,
        "formatted_duration": "6:30",
        "full_title": "The Beatles - Blues song 43 - The blue album",
        "artist_id": "a6f77950-abc2-4eb2-9333-b3924feff571",
        "genre_id": "64ca9927-47e3-4f5e-a93e-49e66965ffff"
      }
    ]
  },
  "artists": [
    {
      "id": "a6f77950-abc2-4eb2-9333-b3924feff571",
      "name": "The Beatles",
      "description": "A band from England",
      "spotify_uri": "674bdae07bccddf91a63",
      "song_ids": [
        "ffaf9b7d-9217-45e3-b7c5-fbda9e6fa890",
        "f10677c4-6741-44f8-9a82-0d538f6380b2",
        "0c1b633d-0cfa-42b9-9846-1ba3bc144b67",
        "873d0e6a-039d-47f9-a222-fc2fb0ced5ff",
        "10804177-cf4b-4131-9e07-f51ba5a10034",
        "62584767-9834-4de2-b236-2c32fbb229b7",
        "55003d48-f9eb-4cf8-833f-dd469433f166",
        "29e0c0cd-8f17-4942-9cdf-27d518a74a08",
        "f8f1eb7e-a01e-482e-baed-d9b9134cca9d"
      ]
    }
  ],
  "genres": [
    {
      "id": "64ca9927-47e3-4f5e-a93e-49e66965ffff",
      "name": "Blues"
    }
  ]
}
```

### CODE REVIEW

File: app/models/mixtape.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class Mixtape
+  include ActiveModel::SerializerSupport
+
+  attr_reader :songs
+
+  def initialize(genre:,songs:)
```
Comment (dthyresson) [inline app/models/mixtape.rb]: Minor code style (space):

`initialize(genre:, songs:)`

File: app/services/mixtape_generator.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  def self.perform(genre:)
+    new(genre: genre).perform
+  end
+
+  def perform
+    Mixtape.new(genre: @genre, songs: pick_songs)
+  end
+
+  def pick_songs
```
Comment (dthyresson) [inline app/services/mixtape_generator.rb:25]: I'm not sure that there is a benefit (eagerloading) to have both the `includes` and `joins`.

Can you get away with just the `joins` and have bullet still say that there's no N+1 query?

From a chaining perspective, tend to order the chains:
- includes/joins
- where
- order by

So, in this case, I'd put `order_by_rand` last in the chain.

This is b/c you might do `order_by_rand(5)` which says give me 5 random songs that uses a `limit` and that's typically the last on the query.

We try to match the order that the SQL might generate.

So try a `.to_sql` after this below and you'll see the `order` be added last.

File: app/services/mixtape_generator.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  def initialize(genre:)
+    @genre = genre
+  end
+
+  def self.perform(genre:)
+    new(genre: genre).perform
+  end
+
+  def perform
+    Mixtape.new(genre: @genre, songs: pick_songs)
```
Comment (dthyresson) [inline app/services/mixtape_generator.rb]: Often have a `attr_reader` here for `genre` so don't have to use `@` for the ivar.

File: app/controllers/v1/mixtapes_controller.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+class V1::MixtapesController < ApplicationController
+  def show
+    mixtape = MixtapeGenerator.perform(genre: params[:genre])
+
+    render json: mixtape
```
Comment (dthyresson) [inline app/controllers/v1/mixtapes_controller.rb]: Will you sideload artists? via an include?

Comment (dthyresson) [inline app/controllers/v1/mixtapes_controller.rb]: Genres, as well?

File: app/services/mixtape_generator.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  def pick_songs
+    Song.order_by_rand
+        .includes(:genre, :artist)
+        .joins(:genre, :artist)
+        .where("lower(genres.name) = ?", @genre.downcase)
```
Comment (dthyresson) [inline app/services/mixtape_generator.rb]: So, why not find the `Genre` by its id and do `songs.genre_id = ?. genre.id`?

Ahhhhhh... see I got confused by seeing `genre` and I thought the entity was being passed it, but it isn't. It is `genre_name` as the parameter.

Maybe change the `def self.perform(genre:)` to `def self.perform(genre_name:)` to clarify.

If I see the named argument, `genre` I'd think to pass a fully hydrated `Genre` item.

File: app/services/mixtape_generator.rb (hunk 3 above)
Comment (dthyresson) [inline app/services/mixtape_generator.rb]: Does the `Mixtape` need a `genre` attribute?

Isn't in going to be the unique genre on the song?

A song as you've defined it can only belong to one Genre.

File: app/models/mixtape.rb (hunk 1 above)
Comment (dthyresson) [inline app/models/mixtape.rb]: Could you just use http://api.rubyonrails.org/classes/ActiveModel/Model.html ?

That way can define `genre:,songs:` or just `songs` and an attribute?

Comment (dthyresson) [inline app/models/mixtape.rb]: You might gain `ActiveModel::SerializerSupport` by default (though not 100% on that).

File: app/services/mixtape_generator.rb (hunk 5 above)
Comment (dgrosenblatt) [inline app/services/mixtape_generator.rb]: That makes sense...will change to `genre_name`. I hadn't taken a step back to look at it but you are right that it is misleading since there is a `Genre` class, but also `params[:genre]` which corresponds to `genre.name`.

File: app/services/mixtape_generator.rb (hunk 2 above)
Comment (dgrosenblatt) [inline app/services/mixtape_generator.rb:25]: Need to try out bullet-- I'll look again but I think I remember leaving out `includes` and seeing 50 queries scroll by in my rails server window.

File: app/services/mixtape_generator.rb (hunk 3 above)
Comment (dgrosenblatt) [inline app/services/mixtape_generator.rb]: Oh right-- the names of the genre(s) could just be sideloaded from the list of songs. My line of thinking was that the response might contain the requested genre name in addition to the songs themselves, but I see now that this is not necessary as an attribute of `Mixtape`.

File: app/services/mixtape_generator.rb (hunk 2 above)
Comment (dgrosenblatt) [inline app/services/mixtape_generator.rb]: Nope, I was wrong, just making it more complicated than it needed to be. Still one query without ```includes```.

File: app/services/mixtape_generator.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+  def perform
+    Mixtape.new(songs: pick_songs)
+  end
+
+  def pick_songs
+    Song.includes(:genre, :artist).where('songs.id in (?)', song_ids)
+  end
+
+  def song_ids
```
Comment (dthyresson) [inline app/services/mixtape_generator.rb:29]: Gonna think about this one tonight.

Comment (dthyresson) [inline app/services/mixtape_generator.rb:29]: I think you want to do this is the database as a `GROUP BY songs.id HAVING sum(songs.duration) > ?`, time

But will double check tonight

Comment (hchood) [inline app/services/mixtape_generator.rb:29]: @dthyresson are you trying to find a way to avoid iterating over the songs?

File: spec/models/genre_spec.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
+
+describe Genre do
+  describe 'associations' do
+    it { should have_many :songs }
+  end
+
+  describe 'validations' do
+    it { should validate_presence_of :name }
```
Comment (hchood) [inline spec/models/genre_spec.rb:9]: I think name should probably be unique.  In which case you'd also want a DB constraint.

File: spec/requests/v1/mixtape_request_spec.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  describe 'GET /mixtape' do
+    context 'with one genre specified' do
+      it 'responds with JSON for a mixtape' do
+        get(mixtape_url, { genre: 'blues' }, accept_headers)
+
+        expect(response).to have_http_status :ok
+        expect(response).to match_response_schema :mixtape
+      end
+    end
+    context 'with a genre and time in seconds specified' do
```
Comment (hchood) [inline spec/requests/v1/mixtape_request_spec.rb]: Good to leave blank line b/w context blocks.

File: spec/requests/v1/mixtape_request_spec.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+      it 'responds with JSON for a mixtape' do
+        get(mixtape_url, { genre: 'blues', time: 3600 }, accept_headers)
+
+        expect(response).to have_http_status :ok
+        expect(response).to match_response_schema :mixtape
+      end
+    end
+  end
```
Comment (hchood) [inline spec/requests/v1/mixtape_request_spec.rb:74]: What happens if I pass no params? might want to test that case too. Or the case where I pass an invalid number for time, or where genre doesn't exist.

File: spec/support/api/schemas/mixtape.json [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+              "duration",
+              "formatted_duration",
+              "full_title",
+              "artist_id"
+            ]
+          }
+        }
+      },
+      "required": [
+        "songs"
```
Comment (hchood) [inline spec/support/api/schemas/mixtape.json:63]: I might also include the artists in the JSON.  Otherwise if the client wants to display the artist's name on the screen, they need to make additional calls to get the artists.

File: spec/support/api/schemas/song.json [hunk 11]
```diff
@@ -76,6 +76,10 @@
         "formatted_duration": {
           "id": "http://spotify-apprentice-app.com/song/title",
           "type": "string"
+        },
+        "genre_id": {
+          "id": "http://spotify-apprentice-app.com/song/title",
+          "type": "string"
```
Comment (hchood) [inline spec/support/api/schemas/song.json]: If genre is not required, this should also be able to be null: `["string", "null"]`

Comment (hchood) [inline spec/support/api/schemas/song.json]: Also you want the `id` property to reflect the name of the property.  So instead of:

`"id": "http://spotify-apprentice-app.com/song/title"`

you want something like:

`"id": "http://spotify-apprentice-app.com/song/genre_id"`

File: spec/support/api/schemas/songs.json [hunk 12]
```diff
@@ -78,6 +78,10 @@
           "formatted_duration": {
             "id": "http://spotify-apprentice-app.com/songs/0/duration",
             "type": "string"
+          },
+          "genre_id": {
+            "id": "http://spotify-apprentice-app.com/song/title",
+            "type": "string"
```
Comment (hchood) [inline spec/support/api/schemas/songs.json]: Same as above.  Also ID should be

`"id": "http://spotify-apprentice-app.com/songs/0/genre_id"`

File: spec/support/api/schemas/mixtape.json [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+                "type": "string"
+              },
+              "full_title": {
+                "id": "http://spotify-apprentice-app.com/mixtape/songs/0/full_title",
+                "type": "string"
+              },
+              "artist_id": {
+                "id": "http://spotify-apprentice-app.com/mixtape/songs/0/artist_id",
+                "type": "string"
+              }
```
Comment (hchood) [inline spec/support/api/schemas/mixtape.json:48]: Is `genre_id` returned in this JSON? Doesn't look like it but you might want to add it to the serializer.

Usually when I add a property to a model I also add it to:
- the serializer
- the JSON schema that reference that model
- the factory

Also, now that a song has a genre, you might want to include the serialized genre along with serialized songs.  (Could be separate PR)

File: app/services/mixtape_generator.rb [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    song_ids = []
+    Song.joins(:genre)
+        .where("lower(genres.name) = ?", genre_name.downcase)
+        .order_by_rand.find_each do |song|
+          song_ids << song.id
+          duration += song.duration
+          return song_ids if duration > time
+        end
+  end
+end
```
Comment (hchood) [inline app/services/mixtape_generator.rb:40]: Probably want a spec to test the `perform` method.  The other helper methods could be `private`, which indicates that they're not for public use and that you don't intend to test them.

File: spec/requests/v1/mixtape_request_spec.rb (hunk 9 above)
Comment (hchood) [inline spec/requests/v1/mixtape_request_spec.rb:74]: You could also test these cases just in a test for the `MixtapeGenerator` service, I suppose.  Then here you could test that the `MixtapeGenerator.perform` method is called with the correct params.

Check out rspec-mocks for docs on how to test those situations: 

https://github.com/rspec/rspec-mocks

There's a few ways to do it but I usually do something like 

``` ruby
allow(Service).to receive(:perform).with(genre: 'whatever', duration: 1000)

get(...)

expect(Service).to have_received(:perform).with(genre: 'whatever', duration: 1000)
```
